### ISSUE TITLE
A large potential energy difference between GROMACS-2021.4 and OpenMM-7.6.0

### ISSUE THREAD

**Issue by Dropletsimuli** [openmm/openmm#3620]
Hi, 
I want to use the OPLS-AA force field with TIP4P/ICE water model on OpenMM-7.6.0 to research my problem. So, I used ParmEd to parse the GROMACS .top/.itp file and .gro file and run the pure water system (7200 TIP4P/ICE water under 298K and 1 atm) on OpenMM. 
```
from __future__ import division, print_function
from openmm.app import *
from openmm import *
from openmm.unit import *
from parmed import load_file
from parmed import gromacs
from mdtraj.reporters import XTCReporter


input_name = 'ice'
gro_name = input_name + '.gro'
top_name = input_name + '.top'

gromacs.GROMACS_TOPDIR = '/home/zzx/research/tools/gromacs-2021.4/share/gromacs/top'
gro = load_file('./prod_nh.gro')
top = gromacs.GromacsTopologyFile(top_name)
top.box = gro.box[:]

# Integration Options

nonbonded_method = PME
nonbonded_cutoff = 0.85 * nanometers
dt = 0.002 * picoseconds  # timestep
temperature = 298 * kelvin
friction = 1.0 / picosecond
pressure = 1.01325 * bar
barostatInterval = 25

# Simulation Options and Prepare the Simulation

print('Building system...')
platform = Platform.getPlatformByName('CUDA')
platformProperties = {'Precision': 'single', 'DeviceIndex': '0'}
topology = top.topology
system = top.createSystem(nonbondedMethod=nonbonded_method,
                          nonbondedCutoff=nonbonded_cutoff,
                          constraints=HBonds)

system.addForce(MonteCarloBarostat(pressure, temperature, barostatInterval))
# integrator = LangevinMiddleIntegrator(temperature, friction, dt)
integrator = NoseHooverIntegrator(temperature, friction, dt)
simulation = Simulation(topology, system, integrator, platform, platformProperties)
simulation.context.setPositions(gro.positions)

# Simulate and Production
...
```
Here are the TIP4P/ICE .itp file and .top file:
```
;
; Note the strange order of atoms to make it faster in gromacs.
;
[ moleculetype ]
; molname	nrexcl
water		2

[ atoms ]
; id	at type	res nr 	residu name	at name	cg nr	charge
1       opls_113        1       SOL      O     1       0.00
2       opls_114        1       SOL      H     1       0.5897
3       opls_114        1       SOL      H     1       0.5897
4       opls_115        1       SOL      M     1      -1.1794

#ifndef FLEXIBLE
[constraints]
;i j funct doh  dhh
1       2       1       0.09572
1       3       1       0.09572
2       3       1       0.15139

#else
[ bonds ]
; i	j	funct	length	force.c.
1	2	1	0.09572	502416.0 0.09572	502416.0
1	3	1	0.09572	502416.0 0.09572	502416.0

[ angles ]
; i	j	k	funct	angle	force.c.
2	1	3	1	104.52	628.02	104.52	628.02
#endif

[ exclusions ]
1	2	3	4
2	1	3	4
3	1	2	4
4	1	2	3

; The position of the virtual site is computed as follows:
;
;		O
;  	      
;	    	D
;	  
;	H		H
;
; const = distance (OD) / [ cos (angle(DOH)) 	* distance (OH) ]
;	  0.01577 nm	/ [ cos (52.26 deg)	* 0.09572 nm	]

; Vsite pos x4 = x1 + a*(x2-x1) + b*(x3-x1)

[ virtual_sites3 ]
; Vsite from			funct	a		b
4	1	2	3	1	0.13458335      0.13458335

```
```
#include "oplsaa.ff/forcefield.itp"
#include "./tip4p-ice.itp"

 [system]
 water_mix
 
 [molecules]
 water  7200
```
And I compared the potential energy of the same  system on GROMACS and OpenMM. However, there are large potential energy difference between GROMACS and OpenMM, where the potential energy on GROMACS is **-390409.90 kJ/mol** and OpenMM is **-5244768.70 kJ/mol**, just as the picture shows:
![potential](https://user-images.githubusercontent.com/78062613/169651355-4d2b5982-40dc-4455-92a3-8b478dec4222.png)
Here are all the input files:
[gitub_issue.zip](https://github.com/openmm/openmm/files/8812518/gitub_issue.zip)

What may causes such a large potential energy difference?

Best,
Drep

**Comment by peastman** [comment]
What happens if you use the GromacsGroFile and GromacsTopFile classes in OpenMM instead of using ParmEd?  If you can reproduce this using OpenMM's loaders, we can look into it and see what's going on.  Otherwise, you should open an issue on the ParmEd repo.

**Comment by Dropletsimuli** [comment]
> What happens if you use the GromacsGroFile and GromacsTopFile classes in OpenMM instead of using ParmEd? If you can reproduce this using OpenMM's loaders, we can look into it and see what's going on. Otherwise, you should open an issue on the ParmEd repo.

I also try it with OpenMM's loaders, but  there was an error , `openmm.OpenMMException: CustomNonbondedForce: Multiple exclusions are specified for particles 0 and 1`. It seems OpenMM's loaders can not correctly operate the virtual sites and exclusion for the gromacs .itp file. 

Here are all the input files:
[test_with_openmm_loader.zip](https://github.com/openmm/openmm/files/8820327/test_with_openmm_loader.zip)

**Comment by peastman** [comment]
Thanks!  I found the problem in GromacsTopFile that led to the duplicate exclusions.  The fix is in #3630.  Can you try it and see how it works now?

**Comment by Dropletsimuli** [comment]
> Thanks! I found the problem in GromacsTopFile that led to the duplicate exclusions. The fix is in #3630. Can you try it and see how it works now?

Thanks for your effort! I try it with your fixs and the `Multiple exclusions`  error did not appear. But there was occurd another error, `openmm.OpenMMException: Particle coordinate is NaN`, even when I selected the last frame obtained from GROMACS npt and nvt simulation as the initial configuration for OpenMM simulation. The system blew up on OpenMM-7.7.0. 

Here are all the input files:
[test_with_openmm_loader2.zip](https://github.com/openmm/openmm/files/8852510/test_with_openmm_loader2.zip)

**Comment by peastman** [comment]
Thanks!  There seem to be a few different problems going on.  The first was a bug in the code for handling explicit constraints from the file.  I fixed it and added the fix to #3630.

The second problem is that your file uses a type of virtual site that isn't currently supported by GromacsTopFile.  Gromacs has a bunch of different specialized virtual site types for different purposes.  The only one currently supported by the loader is type 2.  This was compounded by the fact that your file refers to it using the name `dummies3`, which apparently is an old name for `virtual_sites3` that Gromacs still supports for backward compatibility.

Supporting all the different virtual site types in Gromacs would be a bit of work, but your file uses only a particularly simple one.  I'll add support for it in a separate PR.

**Comment by peastman** [comment]
The three particle vsite is implemented in #3634.

**Comment by Dropletsimuli** [comment]
> The three particle vsite is implemented in #3634.

Thank you! It works.
